cmd/skopeo/sync.go

```diff
@@ -368,61 +378,154 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
+			}
+		}
+
+		if len(filteredSourceReferences) == 0 {
+			logger.Warnf("No refs to sync found")
 			continue
 		}
 
-		repoLogger.Infof("Start filtering using the regular expression: %v", tagRegex)
-		for _, sReference := range allSourceReferences {
-			tagged, isTagged := sReference.DockerReference().(reference.Tagged)
+		repoDescList = append(repoDescList, repoDescriptor{
+			ImageRefs: filteredSourceReferences,
+			Context:   sys,
+		})
+	}
+	return repoDescList
+}
+
+// tagRegexFilterCollection converts a map of (repository name, tag regex) pairs
+// into a filterCollection, which is a map of (repository name, filter function)
+// pairs.
+func tagRegexFilterCollection(collection map[string]string) (filterCollection, error) {
+	filters := filterCollection{}
+
+	for repoName, tagRegex := range collection {
+		pattern, err := regexp.Compile(tagRegex)
+		if err != nil {
+			return nil, err
+		}
+
+		f := func(logger *logrus.Entry, sourceReference types.ImageReference) bool {
+			logger.Infof("Start filtering using the regular expression: %v", tagRegex)
+			tagged, isTagged := sourceReference.DockerReference().(reference.Tagged)
 			if !isTagged {
-				repoLogger.Errorf("Internal error, reference %s does not have a tag, skipping", sReference.DockerReference())
-				continue
+				logger.Errorf("Internal error, reference %s does not have a tag, skipping", sourceReference.DockerReference())
+				return false
 			}
-			if tagReg.MatchString(tagged.Tag()) {
-				sourceReferences = append(sourceReferences, sReference)
+			return pattern.MatchString(tagged.Tag())
+		}
+		filters[repoName] = f
+	}
+
+	return filters, nil
+}
+
+// semverFilterCollection converts a map of (repository name, array of semver constraints) pairs
+// into a filterCollection, which is a map of (repository name, filter function)
+// pairs.
+func semverFilterCollection(collection map[string][]string) (filterCollection, error) {
+	filters := filterCollection{}
+
+	for repoName, constraintStringList := range collection {
+		var constraints []*semver.Constraints
+		for _, constraintString := range constraintStringList {
+			constraint, err := semver.NewConstraint(constraintString)
+			if err != nil {
+				return nil, err
 			}
+			constraints = append(constraints, constraint)
 		}
 
-		if len(sourceReferences) == 0 {
-			repoLogger.Warnf("No refs to sync found")
-			continue
+		f := func(logger *logrus.Entry, sourceReference types.ImageReference) bool {
+			tagged, isTagged := sourceReference.DockerReference().(reference.Tagged)
+			if !isTagged {
+				logger.Errorf("Internal error, reference %s does not have a tag, skipping", sourceReference.DockerReference())
+				return false
+			}
+			tagVersion, err := semver.NewVersion(tagged.Tag())
+			if err != nil {
+				logger.Tracef("Tag %q cannot be parsed as semver, skipping", tagged.Tag())
+				return false
+			}
+			for _, constraint := range constraints {
+				if !constraint.Check(tagVersion) {
+					return false
+				}
+			}
```

Comment:
The previous version accepted a tag if any of the constraints matched; this one only accepts a tag if all of them match. That doesn’t match the documentation, AFAICS for the
```
            - "3.12 - 3.13"
            - ">= 3.17"
```
inputs no tag could ever match.

Reply:
honestly I changed that on purpose I think the relation between array of tags should be "AND" instead of "OR"
let's take this example
```
docker.io:
  images-by-semver:
    node: 
      - "18 - 20"
      - ">= 18.19"
      - "<= 20.0"
```
In the previous code, the selected images were all of these constraints. It even selected old versions from `8` to the newer versions like `21`. In my view, this is confusing because what is the point of having multiple constraints then? If we can't limit the filter more by adding more constraints, otherwise we just need to provide one constraint instead of a list. 
the current setup will reflect the meaning of filter and each constraint in the list will limit the output more 
I'm happy to update the docs to match the current behaviour if we agree on it.

Reply:
Most importantly the code must match the documentation.

---

https://github.com/Masterminds/semver#basic-comparisons shows that there already is a way to AND / OR constraints; there is only one level of nesting, but that should be sufficient to express any policy. (`a AND (b OR c)` == `(a AND b) OR (a AND c)`.

So, in that sense, it doesn’t matter whether we OR or AND the top-level entries — we might not even need multiple lines, just one constraint per repo, and let the user do the boolean math to form it.

I’d argue for the previous (OR) semantics:
- It is consistent with the way `cfg.Images` works; those are ORed
- I don’t expect constraints to be that complicated? The three constraints in your example seem just to be redundant, equivalent to `18.19 - 20.0`; users would presumably just write that one. I would _guess_ that a typical rule to sync an image can well fit on one line, using the `Masterminds/semver` AND/OR language. I don’t have extensive practical experience, that might be wrong.
- OTOH it does seem to me that it would be _fairly frequent_ to sync several narrowly-specified stable versions: “we are deploying foo v1.2, v1.4, and v1.8 throughout the company; grab all 1.X.Y versions of those”. The OR semantics makes it easy to express that.

So I’d prefer OR but that’s not a blocker; the code matching the documentation is a blocker.

Reply:
I think having a single constraint is enough to use and clear from a user perspective 
if the users need more advanced filtering, they can use regex. The point of this feature is to provide an easy way to reference tags 
and single constraint will match the behaviour of regex 

```
docker.io:
  images-by-semver:
    alpine: ">= 3.12"
```

Reply:
Regex and version comparisons seem rather different to me, one is hardly a replacement for the other.

\*shrug\* there is `||` to express OR, so a single expression should work.